firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,828 @@
+    ptr_to_buffer_offset->TFI        = kMessage->TFI;
+    ptr_to_buffer_offset->id         = kMessage->id;
+    ptr_to_buffer_offset->data.qword = kMessage->data.qword;
+    can_registers[controller]->CMR   = command;
+    RegisterIsr(CAN_IRQn, &CanIrqHandler, true, 5);
+  }
+
+  [[gnu::always_inline]] static void ReadMessageFromBuffer(
+      Controllers controller)
+  {
+    RxMessage_t message;
+    memset(&message, 0, sizeof(RxMessage_t));
+
+    volatile RxMessage_t * ptr_to_buffer_offset =
+        reinterpret_cast<volatile RxMessage_t *>(
+            &can_registers[controller]->RFS);
+
+    NVIC_DisableIRQ(CAN_IRQn);
+    message.RFS        = ptr_to_buffer_offset->RFS;
+    message.id         = ptr_to_buffer_offset->id;
+    message.data.qword = ptr_to_buffer_offset->data.qword;
+    NVIC_EnableIRQ(CAN_IRQn);
+
+    xQueueSendFromISR(receive_queue[controller], &message, 0);
+    can_registers[controller]->CMR = kReleaseRxBuffer;
+  }
+
+  [[gnu::always_inline]] void CaptureRegisterData()
+  {
+    interrupts[controller_].ICR    = can_registers[controller_]->ICR;
+    status[controller_].SR         = can_registers[controller_]->SR;
+    global_status[controller_].GSR = can_registers[controller_]->GSR;
+  }
+
+  void SetControllerMode(const Modes kModeBitPos, bool enable_mode)
+  {
+    if (enable_mode == true)
+    {
+      can_registers[controller_]->MOD |= (1 << kModeBitPos);
+    }
+    else
+    {
+      can_registers[controller_]->MOD &= ~(1 << kModeBitPos);
+    }
+  }
+
+  void EnablePower()
+  {
+    // PCONP - Power Control Peripherals Register
+    if (controller_ == kCan1)
+    {
+      PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kCan1);
+    }
+    else
+    {
+      PowerUpPeripheral(Lpc40xxSystemController::Peripherals::kCan2);
+    }
+  }
+
+  void ConfigurePins()
+  {
+    // Configure internal pin MUX to map board I/O pins to controller
+    //                       _
+    //                      / |<--x-->[GPIOx]
+    //                     /  |<--x-->[SPIx]
+    // [Board I/O Pin]<-->|Mux|<----->[CANx]
+    //                     \  |<--x-->[UARTx]
+    //                      \_|<--x-->[I2Cx]
+    //
+    if (controller_ == kCan1)
+    {
+      rd_->SetPinFunction(kRd1FunctionBit);
+      td_->SetPinFunction(kTd1FunctionBit);
+    }
+    else
+    {
+      rd_->SetPinFunction(kRd2FunctionBit);
+      td_->SetPinFunction(kTd2FunctionBit);
+    }
+  }
+
+  void EnableInterrupts()
+  {
+    // IER - Interrupt Enable Register
+    can_registers[controller_]->IER |= (1 << kRxBufferIntEnableBit);
+    can_registers[controller_]->IER |= (1 << kTxBuffer1IntEnableBit);
+    can_registers[controller_]->IER |= (1 << kTxBuffer2IntEnableBit);
+    can_registers[controller_]->IER |= (1 << kTxBuffer3IntEnableBit);
+  }
+
+  // TODO(rabeelio): Look into dynamically calculating baud rates for different
+  // peripheral bus clock speeds.
+  // TODO(rabeelio): Add details about these calculations in documentation.
```

Comment:
Add two more TODO(s)
1. Make this public, so the user can change the baud rate if they wish
2. This function should NOT assume the bus rate is 48MHz. Use the GetPeripheralFrequency() from system controller, which you have inheritated, to get the actual clock and do the calculation to generate baud rate for the can bus.

Reply:
1. hmm, but the user can just set the baud rate when the call the class constructor, right? Or is better to have both?
2. Right, forgot that existed. Will add it.